### DISCUSSION TITLE
Selectors vs Functions

### DISCUSSION THREAD

**Discussion by paulmccrady** [Azure/PSRule.Rules.Azure#2521] [Q&A] [answered]
In something like Azure.Template.Rule.ps1 there are functions defined (e.g IsParameterFile) which is then used with -If ...
What is the difference with this approach and using a Selector (-With) ?
https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Selectors/

**Comment by BernieWhite** [comment, marked as answer]
@paulmccrady When writing PowerShell rules you can use either. When using YAML/ JSON rules you are only able to use a Selector.

In general, selectors (`-With`) and script pre-conditions (`-If`) can be used interchangeably.

We prefer use of Selectors over script pre-conditions because we are making efforts to optimize this use case. However, currently not all expressions or use cases can be supported using the YAML/ JSON syntax supported by selectors today.
